src-java/kilda-pce/src/main/java/org/openkilda/pce/impl/InMemoryPathComputer.java

```diff
@@ -223,48 +236,116 @@ private WeightFunction getWeightFunctionByStrategy(PathComputationStrategy strat
+
+        if (edge.getDestSwitch().getDiversityGroupUseCounter() > 0) {
+            long value = edge.getDestSwitch().getDiversityGroupUseCounter() * config.getDiversitySwitchLatency();
+            if (isForProtectedPath) {
+                penalties.put(PROTECTED_DIVERSITY_SWITCH_LATENCY, value);
+            } else {
+                penalties.put(DIVERSITY_SWITCH_LATENCY, value);
+            }
+        }
+
+        if (edge.getAffinityGroupUseCounter() > 0) {
+            long value = edge.getAffinityGroupUseCounter() * config.getAffinityIslLatency();
+            penalties.put(AFFINITY_ISL_LATENCY, value);
+        }
+
+        long edgeLatency = edge.getLatency() <= 0 ? config.getDefaultIslLatency() : edge.getLatency();
+        return new PathWeight(edgeLatency, penalties);
     }
 
     private PathWeight weightByCostAndAvailableBandwidth(Edge edge) {
-        long total = edge.getCost() == 0 ? config.getDefaultIslCost() : edge.getCost();
+        Map<Penalty, Long> penalties = new EnumMap<>(Penalty.class);
+
         if (edge.isUnderMaintenance()) {
-            total += config.getUnderMaintenanceCostRaise();
+            penalties.put(UNDER_MAINTENANCE, (long) config.getUnderMaintenanceCostRaise());
         }
+
         if (edge.isUnstable()) {
-            total += config.getUnstableCostRaise();
+            penalties.put(UNSTABLE, (long) config.getUnstableCostRaise());
         }
-        total += edge.getDiversityGroupUseCounter() * config.getDiversityIslCost()
-                + edge.getDiversityGroupPerPopUseCounter() * config.getDiversityPopIslCost()
-                + edge.getDestSwitch().getDiversityGroupUseCounter() * config.getDiversitySwitchCost()
-                + edge.getAffinityGroupUseCounter() * config.getAffinityIslCost();
-        return new PathWeight(total, edge.getAvailableBandwidth());
+
+        if (edge.getDiversityGroupUseCounter() > 0) {
+            int value = edge.getDiversityGroupUseCounter() * config.getDiversityIslCost();
+            penalties.put(DIVERSITY_ISL_LATENCY, (long) value);
+        }
+
+        if (edge.getDiversityGroupPerPopUseCounter() > 0) {
+            int value = edge.getDiversityGroupPerPopUseCounter() * config.getDiversityPopIslCost();
+            penalties.put(DIVERSITY_POP_ISL_COST, (long) value);
+        }
+
+        if (edge.getDestSwitch().getDiversityGroupUseCounter() > 0) {
+            int value = edge.getDestSwitch().getDiversityGroupUseCounter() * config.getDiversitySwitchCost();
+            penalties.put(DIVERSITY_SWITCH_LATENCY, (long) value);
+        }
+
+        if (edge.getAffinityGroupUseCounter() > 0) {
+            long value = edge.getAffinityGroupUseCounter() * config.getAffinityIslCost();
+            penalties.put(AFFINITY_ISL_LATENCY, value);
+        }
+
+        long cost = edge.getCost() == 0 ? config.getDefaultIslCost() : edge.getCost();
+        return new PathWeight(cost, penalties, edge.getAvailableBandwidth());
```

Comment:
we have a lot all duplicate code in weight functions. Maybe we should move common penalties to some common method?

Reply:
actually no.